src/semidiscretization_euler_gravity.jl

```diff
@@ -0,0 +1,304 @@
+  println(io, "SemidiscretizationEulerGravity using")
+  print(io, "  "); show(io, mime, semi.semi_euler); println()
+  print(io, "  "); show(io, mime, semi.semi_gravity); println()
+  print(io, "  "); show(io, mime, semi.parameters); println()
+  print(io, "  cache with fields:")
+  for key in keys(semi.cache)
+    print(io, " ", key)
+  end
+end
+
+
+@inline function mesh_equations_solver_cache(semi::SemidiscretizationEulerGravity)
+  mesh_equations_solver_cache(semi.semi_euler)
+end
+
+
+@inline function compute_coefficients(t, semi::SemidiscretizationEulerGravity)
+  compute_coefficients!(semi.cache.u_ode, t, semi.semi_gravity)
+  compute_coefficients(t, semi.semi_euler)
+end
+
+@inline function compute_coefficients!(u_ode, t, semi::SemidiscretizationEulerGravity)
+  compute_coefficients!(semi.cache.u_ode, t, semi.semi_gravity)
+  compute_coefficients!(u_ode, t, semi.semi_euler)
+end
+
+
+@inline function calc_error_norms(func, u, t, analyzer, semi::SemidiscretizationEulerGravity)
+  calc_error_norms(func, u, t, analyzer, semi.semi_euler)
+end
+
+
+function rhs!(du_ode, u_ode, semi::SemidiscretizationEulerGravity, t)
+  @unpack semi_euler, semi_gravity, cache = semi
+
+  u_euler   = wrap_array(u_ode , semi_euler)
+  du_euler  = wrap_array(du_ode, semi_euler)
+  u_gravity = wrap_array(cache.u_ode, semi_gravity)
+
+  time_start = time_ns()
+
+  # standard semidiscretization of the compressible Euler equations
+  @timeit_debug timer() "Euler solver" rhs!(du_ode, u_ode, semi_euler, t)
+
+  # compute gravitational potential and forces
+  @timeit_debug timer() "gravity solver" update_gravity!(semi, u_ode)
+
+  # add gravitational source source_terms to the Euler part
+  if ndims(semi_euler) == 2
+    @views @. du_euler[2, .., :] -=  u_euler[1, .., :] * u_gravity[2, .., :]
+    @views @. du_euler[3, .., :] -=  u_euler[1, .., :] * u_gravity[3, .., :]
+    @views @. du_euler[4, .., :] -= (u_euler[2, .., :] * u_gravity[2, .., :] +
+                                     u_euler[3, .., :] * u_gravity[3, .., :])
+  elseif ndims(semi_euler) == 3
+    @views @. du_euler[2, .., :] -=  u_euler[1, .., :] * u_gravity[2, .., :]
+    @views @. du_euler[3, .., :] -=  u_euler[1, .., :] * u_gravity[3, .., :]
+    @views @. du_euler[4, .., :] -=  u_euler[1, .., :] * u_gravity[4, .., :]
+    @views @. du_euler[5, .., :] -= (u_euler[2, .., :] * u_gravity[2, .., :] +
+                                     u_euler[3, .., :] * u_gravity[3, .., :] +
+                                     u_euler[4, .., :] * u_gravity[4, .., :])
+  else
+    error("Number of dimensions $(ndims(semi_euler)) not supported.")
+  end
+
+  runtime = time_ns() - time_start
+  put!(semi.performance_counter, runtime)
+
+  return nothing
+end
+
+
+# TODO: Taal refactor, add some callbacks or so within the gravity update to allow investigating/optimizing it
+function update_gravity!(semi::SemidiscretizationEulerGravity, u_ode::AbstractVector)
```

Comment:
It would be possible in principle. However, we would need to add our optimized methods to the DiffEq ecosystem. For simplicity right now, I decided to keep the previous structure. I don't expect to gain much from moving to DiffEq at this point but it should be possible.

Reply:
OK, fair enough!